File: ubuntu/WallpaperInfoApp/WallpaperServiceInfo.py
#!/usr/bin/env python3

import ThemeInfo as t
import PlatformPreference as pref
import WallpaperService

class Singleton(type):
    _instances = {}
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            cls._instances[cls] = super(Singleton, cls).__call__(*args, **kwargs)
        return cls._instances[cls]

class WallpaperServiceInfo(metaclass=Singleton):

    def __init__(self):
        pref.getPreference(self)
        # must update custom info before it updates themeinfo
        t.setCustomConfigDetail(self._customThemeInfo._root, self._customThemeInfo._allow,  self._customThemeInfo._filter)
        self._themeInfo  = t.getThemeInfo(self._theme)
        self._wallpaperService = WallpaperService.WallpaperService()
        self._currentWPath = None
        self._thumbnail = None

    # persistent property setter/getter 
    def getSourceRootPath(self):
        return self._sourceRootPath
    def setSourceRootPath(self, path):
        self._sourceRootPath = path
        pref.setPreference("root_path", path)

    def getTheme(self):
        return self._themeInfo._theme
    def getThemeInfo(self):
        return self._themeInfo
    def setThemeInfo(self, themeInfo):
        self._themeInfo = themeInfo
        pref.setPreference("theme", str(int(self._themeInfo._theme)))
    def setCustomThemeInfo(self, customThemeInfo):
        self._customThemeInfo = customThemeInfo
        pref.setPreference("custom_root", customThemeInfo._root)
        pref.setPreference("custom_allow", customThemeInfo._allow)
        pref.setPreference("custom_filter", customThemeInfo._filter)
    def getCustomConfigString(self):
        return self._customThemeInfo._root + ";" + self._customThemeInfo._allow + ";" + self._customThemeInfo._filter
        
    def getPause(self):
        return self._pause
    def setPause(self, pause):
        self._pause = pause
        pref.setPreference("pause", str(pause))

    def getInterval(self):
        return self._interval
    def setInterval(self, interval):
        self._interval = interval
        pref.setPreference("interval", str(interval))
    
    # runtime property setter/getter
    def getcurrentWPath(self):
        return self._currentWPath
    def setcurrentWPath(self, wpath):
        self._currentWPath = wpath

    def getThumbnail(self):
        return self._thumbnail
    def setThumbnail(self, thumbnail):
        if thumbnail != self._thumbnail and self._thumbnail != None:
            del self._thumbnail
        self._thumbnail = thumbnail

    def getLastUsedPaths(self):
        return self._lastUsedPaths
    def setLastUsedPaths(self, lastUsedPaths):
        self._lastUsedPaths = lastUsedPaths

    def getWallpaperService(self):
        return self._wallpaperService

    # public value from resource
    _defaultInterval = 7
    _minInterval = 5 
    _maxInterval = 30

#wpsi = WallpaperServiceInfo()
#print(wpsi.getTheme().label())
#wpsi3 = WallpaperServiceInfo()
#print(wpsi3.getTheme().label())
#wpsi.setThemeInfo(t.getThemeInfo(t.Theme.special1))
#print(wpsi.getTheme().label())
#wpsi2 = WallpaperServiceInfo()
#print(wpsi2.getTheme().label())
#print(wpsi3.getTheme().label())
#wpsi.setInterval(13)

File: apps/mobile-android/wallpaperinfo/app/src/main/java/com/thirdwavesoft/wallpaperinfo/WallpaperServiceInfo.java
package com.thirdwavesoft.wallpaperinfo;

import android.graphics.Bitmap;
import android.content.Context;

import android.util.Log;

public class WallpaperServiceInfo {
	
    private static class SingletonHolder {
        public static final WallpaperServiceInfo instance = new WallpaperServiceInfo();
    }

    public static WallpaperServiceInfo getInstance() {
        return SingletonHolder.instance;
    }

    synchronized private void initPropertiesFromPreference() {
        _minInterval = _appContext.getResources().getInteger(R.integer.min_interval);
        _maxInterval = _appContext.getResources().getInteger(R.integer.max_interval);
        _defaultInterval = _appContext.getResources().getInteger(R.integer.default_interval);

        WallpaperInfoPreference wpref = PlatformPreference.getPreferences(_appContext);
        _sourceRootPath = wpref.root_path;
        _customThemeInfo._root = wpref.custom_root;
        _customThemeInfo._allow = wpref.custom_allow;
        _customThemeInfo._filter = wpref.custom_filter;
        // must update custom info before it updates themeinfo
        ThemeInfo.setCustomConfig(_customThemeInfo._root, _customThemeInfo._allow, _customThemeInfo._filter);
        _themeInfo = ThemeInfo.getThemeInfo(wpref.theme);
        _pause = wpref.pause;
        _interval = wpref.interval;
        _saverTime = wpref.saverTime;
        if (_interval < _minInterval || _interval > _maxInterval) {
            _interval = _defaultInterval;
        }
        if (_sourceRootPath == null || !BPUtil.fileExists(_sourceRootPath)) {
            _sourceRootPath = ImageFileManager.getDefaultSourceRootPath();
        }
    }

    // persistent property setter/getter
    synchronized public String getSourceRootPath() {
        return _sourceRootPath;
    }
    synchronized public void setSourceRootPath(String path) {
        _sourceRootPath = path;
        PlatformPreference.setPreference(_appContext, "root_path", path);
	}

    synchronized public Theme getTheme() { return _themeInfo._theme; }
    synchronized public ThemeInfo getThemeInfo() { return _themeInfo; }
    synchronized public void setThemeInfo(ThemeInfo themeInfo) {
        _themeInfo = themeInfo;
        PlatformPreference.setPreference(_appContext, "theme", (new Integer(themeInfo._theme.intValue())).toString());
    }
    synchronized public void setCustomThemeInfo(ThemeInfo customThemeInfo) {
        _customThemeInfo = customThemeInfo;
        PlatformPreference.setPreference(_appContext, "custom_root", customThemeInfo._root);
        PlatformPreference.setPreference(_appContext, "custom_allow", customThemeInfo._allow);
        PlatformPreference.setPreference(_appContext, "custom_filter", customThemeInfo._filter);
    }
    synchronized public String getCustomConfigString() {
        return _customThemeInfo._root + ";" + _customThemeInfo._allow + ";" + _customThemeInfo._filter;
    }

    synchronized public boolean getPause() {
        return _pause;
    }
    synchronized public void setPause(boolean pause) {
    	_pause = pause;
        PlatformPreference.setPreference(_appContext, "pause", (new Boolean(pause)).toString());
    }

    synchronized public int getInterval() {
	    return _interval;
	}
    synchronized public void setInterval(int interval) {
	    _interval = interval;
        PlatformPreference.setPreference(_appContext,"interval", (new Integer(interval)).toString());
	}

    // runtime property setter/getter
    synchronized public Context getAppContext() {
        return _appContext;
    }
    synchronized public void setAppContext(Context context) {

        if (_appContext == null) {
            _appContext = context.getApplicationContext();
            if (_appContext != null) {
                initPropertiesFromPreference();
                BPUtil.BPLog("%s (%s)", "WallpaperServiceInfo Preference ", _appContext.getApplicationInfo().toString());
            } else {
                Log.e(_TAG, "WallpaperServiceInfo reading preference failed due to no conext.");
            }
        } else if (_appContext != context.getApplicationContext()) {
            Log.e(_TAG, "WallpaperServiceInfo context is changed.");
        }
    }

    synchronized public WPath getcurrentWPath() {
        return _currentWPath;
    }
    synchronized public void setcurrentWPath(WPath path) {
        _currentWPath = path;
    }

    synchronized public Bitmap getThumbnail() {
        return _thumbnail;
    }
    synchronized public void setThumbnail(Bitmap thumbnail) {
        if (thumbnail == null && _thumbnail != null) {
            _thumbnail.recycle();
        }
        _thumbnail = thumbnail;
    }

    synchronized public WLinkedHashMap<String, String> getLastUsedPaths() {
        return _lastUsedPaths;
    }
    synchronized public void setLastUsedPaths(WLinkedHashMap<String, String> lastUsedPaths) {
        _lastUsedPaths = lastUsedPaths;
    }

    synchronized public ServiceMode getMode() {
        return _mode;
    }
    synchronized public void setMode(ServiceMode mode) {
        _mode = mode;
    }

    synchronized public boolean getSaver() {
        return _saver;
    }
    synchronized public void setSaver(boolean saver) {
        _saver = saver;
    }

    synchronized public int getSaverTime() {
        return _saverTime;
    }
    synchronized public void setSaverTime(int saverTime) {
        _saverTime = saverTime;
        //PlatformPreference.setPreference(_appContext,"saver_time", (new Integer(_saverTime)).toString());
    }


    synchronized public boolean isActivityonTop() {
        return _activityonTop;
    }
    synchronized public void setActivityonTop(boolean activityonTop) {
        _activityonTop = activityonTop;
    }

    // persistent property
    private String _sourceRootPath = ImageFileManager.getDefaultSourceRootPath();
    private ThemeInfo _themeInfo = ThemeInfo.getThemeInfo(Theme.default1);
    private ThemeInfo _customThemeInfo = new ThemeInfo(Theme.custom,
            ThemeInfo._default_custom_root,
            ThemeInfo._default_custom_allow,
            ThemeInfo._default_custom_filter);
	private boolean _pause = false;
    private int _interval = _defaultInterval; // in seconds

    // runtime property
    private WPath _currentWPath = null;
    private Bitmap _thumbnail = null;
    private WLinkedHashMap<String, String> _lastUsedPaths = new WLinkedHashMap<String, String>();
    private ServiceMode _mode = ServiceMode.wallpaper;
    private boolean _saver = false;
    private int _saverTime = _defaultSaverTime;

    // Platform Specifc runtime property
    private Context _appContext = null;
    private boolean _activityonTop = false;

    //public value from resource
    public static int _defaultInterval = 15; // 15 will not be used. just placeholder.
    public static int _defaultSaverTime = 5; // 5 will not be used. just placeholder.
    public static int _minInterval = 5; // 5 will not be used. just placeholder.
    public static int _maxInterval = 30; // 30 will not be used. just placeholder.

    // default values
    private final static String _TAG = "WallpaperServiceInfo";

}

enum ServiceMode {
    wallpaper, slideshow, cast;
    static ServiceMode rawValue(int index) {
        if (index >=0 && index < _serviceModeValues.length) {
            return _serviceModeValues[index];
        }
        return wallpaper;
    }
    private static ServiceMode[] _serviceModeValues = values();
    int intValue() {
        return ordinal();
    }
};

